### PR TITLE
Add simplified bicycle simulation

### PR DESCRIPTION
**Pull request by oliverlee**

### CODE REVIEW

**Comment by oliverlee** [comment]
@mickvangelderen Do you want to review this or give your thoughts?

Specifically, I'm wondering if you know a better way to "override" the `output_index_t` enum class defined in the `Bicycle` base class. In the derived class `BicycleArend`, I define enum class `output_index_t` which shadows the definition in the base class.

I want to have setter and getter functions for this new enum class but I defined:
 `void set_output_element(output_t&, output_index_t, real_t)` and
 `real_t get_output_element(const output_t&, output_index_t)`
as static in `Bicycle` so they cannot be overriden. Instead I define a member function variants for `BicycleArend` to be used instead.

If the `Bicycle` setter/getter is used with `BicycleArend::output_index_t`, the compiler will throw an error.

File: src/bicycle/arend.cc [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    return x[index(field)];
+}
+
+BicycleArend::output_t BicycleArend::normalize_output(const output_t& y) const {
+    static_assert(index(output_index_t::steer_angle) == 0,
+        "Invalid underlying value for output index element");
+    static_assert(index(output_index_t::steer_rate) == 1,
+        "Invalid underlying value for output index element");
+    static_assert(index(output_index_t::number_of_types) == 2,
+        "Invalid underlying value for output index element");
```
Comment (mickvangelderen) [inline src/bicycle/arend.cc:140]: Odd location for these assertions.

File: inc/bicycle/arend.h [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+#include <boost/numeric/odeint/algebra/vector_space_algebra.hpp>
+
+namespace model {
+
+class BicycleArend final : public Bicycle {
+    public:
+        enum class output_index_t: uint8_t {
+            steer_angle = 0,
+            steer_rate,
+            number_of_types
```
Comment (mickvangelderen) [inline inc/bicycle/arend.h:13]: I'd set all enum values explicitly because they matter.

File: src/bicycle/arend.cc (hunk 1 above)
Comment (oliverlee) [reply src/bicycle/arend.cc:140]: Where would you put them? I chose that location so I would remember to update that function if the enum class is changed.

**Comment by mickvangelderen** [comment]
> @mickvangelderen Do you want to review this or give your thoughts?

I suddenly realized that setting and getting state elements are dependent on the type of the state. Would it make sense to do something like

```
template magic if (model_t::state_t is ArendState) {
    model->state->set_steer_angle(...);
    // or
    ArendState::set_steer_angle(model->state, ...);
    ...
}
```

**Comment by oliverlee** [comment]
The state types (`state_t`, `full_state_t`, ...) are inherited from `Bicycle` and are the same in all derived classes. `state_t` is also a subset of `full_state_t` so you would still need to explicitly specify what state type you want to set/get.

The model classed don't hold any states as well, only the observer classes do.

**Comment by mickvangelderen** [comment]
If what you have works right now that will have to do. I need a lot of time to understand why things are implemented the way they are which I do not have right now.

**Comment by oliverlee** [comment]
We can come back to it later if it turns out to be unclear or problematic.

**Pull request merged**